### PR TITLE
openZeppelin upgradeable 

### PR DESCRIPTION
**Pull request by i-stam**
- Uses the `upgradeability` package offered by openZeppelin

- Uses one external `initialize()` function for the wallet with the appropriate `initializer` modifier, the rest of the `initialize()` functions are internal and manual care should be taken for invoking them.

- The rest of the contract code was left untouched (initializing contract parameters e.g. ENS nodes, should be done in a separate PR)

- The openZeppelin contracts are considered safe by default and no additional testing is performed. We just tested if they were initialized and used properly and slightly modified them to allow the admin (aka wallet owner) to delegate calls.  Repo commit hash:9baca3afb5649b6defc3a75eeb69f4930852180f (https://github.com/OpenZeppelin/openzeppelin-sdk/commit/9baca3afb5649b6defc3a75eeb69f4930852180f)

### CODE REVIEW

File: contracts/controller.sol [hunk 1]
```diff
@@ -57,7 +57,9 @@ contract Controller is IController, Ownable, Transferrable {
 
     /// @notice Constructor initializes the owner with the provided address.
     /// @param _ownerAddress_ address of the owner.
-    constructor(address payable _ownerAddress_) public Ownable(_ownerAddress_, false) {}
+    constructor(address payable _ownerAddress_) public {
```
Comment (mischat) [inline contracts/controller.sol:60]: I guess we don't need to make the rest of it upgradable as we are using ens anyways.

Comment (i-stam) [reply contracts/controller.sol:60]: Yes, making the wallet upgradeable is the main goal, we don't really care about the rest.

File: contracts/holder.sol [hunk 2]
```diff
@@ -50,12 +50,10 @@ contract Holder is Balanceable, ENSResolvable, Controllable, Transferrable, Toke
     /// @param _ens_ is the address of the ENS registry.
     /// @param _tokenWhitelistNode_ is the ENS node of the Token whitelist.
     /// @param _controllerNode_ is the ENS node of the Controller
-    constructor(address _burnerContract_, address _ens_, bytes32 _tokenWhitelistNode_, bytes32 _controllerNode_)
-        public
-        ENSResolvable(_ens_)
-        Controllable(_controllerNode_)
-        TokenWhitelistable(_tokenWhitelistNode_)
-    {
+    constructor(address _burnerContract_, address _ens_, bytes32 _tokenWhitelistNode_, bytes32 _controllerNode_) public {
```
Comment (nostdm) [inline contracts/holder.sol:53]: Why did we remove the constructors from all contracts instead of just adding the initialize() interface to the wallet?
I would imagine it still makes sense to invoke regular constructors() for contracts that don't need to be upgradable.

File: contracts/internals/controllable.sol [hunk 3]
```diff
@@ -58,6 +48,16 @@ contract Controllable is ENSResolvable {
         return _controllerNode;
     }
 
+    /// @notice Initializes the controller contract object.
+    /// @param _controllerNode_ is the ENS node of the Controller.
+    /// @dev pass in bytes32(0) to use the default, production node labels for ENS
+    function _initializeControllable(bytes32 _controllerNode_) internal {
```
Comment (nostdm) [inline contracts/internals/controllable.sol:54]: I made changes in the controller v2 branch where I added default values to ENS variables.
IMO the _controllerNode should have a default value of controller.tokencard.eth and then we could remove the constructor and just have a separate method to setControllerNode() otherwise this intialize function is not safe because if you forget to invoke it in the constructor you will deploy a broken contract. Which isn't the case with a constructor because the compiler will tell you about it.

File: contracts/internals/ensResolvable.sol [hunk 4]
```diff
@@ -28,9 +28,10 @@ contract ENSResolvable {
     /// @notice _ensRegistry points to the ENS registry smart contract.
     address private _ensRegistry;
 
-    /// @param _ensReg_ is the ENS registry used
-    constructor(address _ensReg_) internal {
-        _ensRegistry = _ensReg_;
+    /// @notice Checks if the contract has been initialized succesfully i.e. the ENS registry has been set.
+    modifier initialized() {
```
Comment (nostdm) [inline contracts/internals/ensResolvable.sol:32]: I think we should keep the constructors or do what I did in the controller v2 branch which is to remove the constructors and set a default value for _ensRegistry

Comment (i-stam) [reply contracts/internals/ensResolvable.sol:32]: All these comments are good point and as I wrote before I found that there is a bit of inconsistency on the way we initialize things.

File: contracts/controller.sol (hunk 1 above)
Comment (mischat) [reply contracts/controller.sol:60]: great!

File: contracts/externals/initializable.sol [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+/**
+ * @title Initializable
+ *
+ * @dev Helper contract to support initializer functions. To use it, replace
+ * the constructor with a function that has the `initializer` modifier.
+ * WARNING: Unlike constructors, initializer functions must be manually
+ * invoked. This applies both to deploying an Initializable contract, as well
+ * as extending an Initializable contract via inheritance.
+ * WARNING: When used with inheritance, manual care must be taken to not invoke
+ * a parent initializer twice, or ensure that all initializers are idempotent,
```
[the remark this answers is not in the document]
Comment (mischat) [reply contracts/externals/initializable.sol:13]: I guess this is something we should get to the bottom of. Well spotted. 

Naively i would say yes, they are, but worth going through it all to confirm.

File: contracts/holder.sol (hunk 2 above)
Comment (mischat) [reply contracts/holder.sol:53]: +1

File: contracts/internals/controllable.sol (hunk 3 above)
Comment (mischat) [reply contracts/internals/controllable.sol:54]: I like that pattern, this is something that has caught us out before.

File: contracts/externals/initializable.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (i-stam) [reply contracts/externals/initializable.sol:13]: So the pattern we are following is one external initializer (wallet) and the rest are internal (my assumption is that they can never be called again unless it is explicitly done via the inheriting contract.

File: contracts/externals/initializable.sol [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    // under construction or not.
+    address self = address(this);
+    uint256 cs;
+    assembly { cs := extcodesize(self) }
+    return cs == 0;
+  }
+
+  // Reserved storage space to allow for layout changes in the future.
+  uint256[50] private ______gap;
+}
```
[the remark this answers is not in the document]
Comment (i-stam) [reply contracts/externals/initializable.sol:62]: I guess we can add it even though it is an external contract :)

File: contracts/holder.sol (hunk 2 above)
Comment (i-stam) [reply contracts/holder.sol:53]: The `constructor` wasn't removed in independent contracts like holder above, they are not _upgradeable_.

File: contracts/internals/controllable.sol (hunk 3 above)
Comment (i-stam) [reply contracts/internals/controllable.sol:54]: Specifically for the controllable, there is a constant assigned during initialization. I left the `If` statement there even though I prefer a `require` one because it messes up with the unit tests and the tool suites (call me echidna). It was done in this way, so that wallet deployed can pass a 0 value and no problem would arise.

File: contracts/internals/ensResolvable.sol (hunk 4 above)
Comment (i-stam) [reply contracts/internals/ensResolvable.sol:32]: The constructors for inherited contracts can indeed be kept (even though the guideline is to avoid them for upgradeable contracts) It is just code replication and in the end, you always have to invoke them (in many cases we forgot to invoke them and our tests were failing and it was hard to debug them). If it is just for aesthetic reasons, my guess is that we can keep them.

**Comment by i-stam** [comment]
> I would assume this upgradable proxy could be used with the 'library' approach that we spoke about few days ago - and could also be simplified - since at the moment we are adding more contracts on top of the already complex wallet contract.

If there are test that are missing, we should add them , feel free to suggest. All the previous tests are running successfully and just the basic tests for the upgradeable contracts are added. But this was supposed to be fast so we can send it to the auditor, we can add more tests in the meanwhile.

> P.S. Can we make the `externals/upgradeability` directory called `externals/upgradable` or something cleaner/simpler?
I think there is no problem there, we can change the name to something else. This is actually the name that openZeppeling uses so I kept it the same as we do for all the external contracts

File: contracts/externals/initializable.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (mischat) [reply contracts/externals/initializable.sol:13]: OK, so i have been through these constructors. 

``` ~/go/src/github.com/tokencard/contracts/contracts/internals   upgradeable-zeppelin ⍟2  sort -u /tmp/lame
_initializeControllable(_controllerNode_);
_initializeENSResolvable(_ens_);
_initializeGasTopUpLimit();
_initializeLoadLimit(_tokenWhitelistNode_);
_initializeOwnable(_ownerAddress_, false);
_initializeOwnable(_owner_, _transferable_);
_initializeSpendLimit(_spendLimit_);
_initializeTokenWhitelistable(_tokenWhitelistNode);
_initializeTokenWhitelistable(_tokenWhitelistNode_);```

I have been through all of those methods to check if they are idempotent :

This initialise method isn’t idempotent, from the Wallet, it has a time based component insofar as “now”:

```
    function _initializeLoadLimit(bytes32 _tokenWhitelistNode) internal {
        _initializeTokenWhitelistable(_tokenWhitelistNode);
        (, uint256 stablecoinMagnitude, , , , , ) = _getStablecoinInfo();
        require(stablecoinMagnitude > 0, "no stablecoin");
        _maximumLoadLimit = _MAXIMUM_STABLECOIN_LOAD_LIMIT * stablecoinMagnitude;
        _loadLimit = DailyLimitTrait.DailyLimit(_maximumLoadLimit, _maximumLoadLimit, now, 0, false);
    }
```

[the remark this answers is not in the document]
Comment (mischat) [reply contracts/externals/initializable.sol:13]: I am guessing initialise functions with "now" in them aren't good, but i will leave you guys to figure this out :)

File: contracts/holder.sol (hunk 2 above)
Comment (mischat) [reply contracts/holder.sol:53]: noted

File: contracts/internals/ensResolvable.sol (hunk 4 above)
Comment (mischat) [reply contracts/internals/ensResolvable.sol:32]: I have looked at these too. I am indifferent to changing them now or in the future, but :

```
I like the default value set in the Controllable.sol :

_initializeControllable


This one is different and I prefer the above way of doing things : 

_initializeENSResolvable```

Comment (i-stam) [reply contracts/internals/ensResolvable.sol:32]: by the way, if we follow this approach then we can even remove the harcoded ens values in the deployer contract. but every time we have to edit separate contracts. I ' fine with that, but let's change in a separate PR and think a bit of what we want.

File: contracts/externals/initializable.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (i-stam) [reply contracts/externals/initializable.sol:13]: I guess the whole confusion comes from the term itself. The comment is :
`When used with inheritance, manual care must be taken to not invoke a parent initializer twice, or ensure that all initializers are idempotent`

I don't exactly understand what _idempotent_ means in this context but my guess is that we shouldn't assume the strict mathematical definition here: if  an initializer can be called again with different parameters and change the state then it shouldn't be considered _idempotent_. That's why I followed the first approach ensuring that the initializers cannot be invoked twice. Please verify my statement, is ti safe to consider that internal functions cannot be invoked in any way unless explicitly stated in the code?

[the remark this answers is not in the document]
Comment (mischat) [reply contracts/externals/initializable.sol:13]: I guess i will leave this to @i-stam and @nostdm to get to the bottom of.

[the remark this answers is not in the document]
Comment (mischat) [reply contracts/externals/initializable.sol:13]: thanks for the note though. All i did is what i knew :)

**Comment by mischat** [comment]
@i-stam the other thing which I would like to get in in this PR is a link to the commit which we took the the open-zepplin upgradebale. Would be good to have the git commit captured

File: contracts/externals/initializable.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (nostdm) [reply contracts/externals/initializable.sol:13]: Non-Idempotent means that by invoking the method with the same parameters it will have different side effects based on whether it's being called the first, second or third time etc.

But I think this is actually handled by the `initializer` modifier from openzeppelin since it prevents the initializer being called multiple times.

File: contracts/internals/ensResolvable.sol (hunk 4 above)
Comment (nostdm) [reply contracts/internals/ensResolvable.sol:32]: Yeah, I've read the openzeppelin docs and I'm OK with the initializer instead of constructor for wallet contracts.

File: contracts/internals/transferrable.sol [hunk 7]
```diff
@@ -34,7 +34,8 @@ contract Transferrable {
     function _safeTransfer(address payable _to, address _asset, uint256 _amount) internal {
         // address(0) is used to denote ETH
         if (_asset == address(0)) {
-            _to.transfer(_amount);
+            (bool success, ) = _to.call.value(_amount)("");
```
Comment (nostdm) [inline contracts/internals/transferrable.sol:37]: Looks good, though what's the reason for this change - gas limit?

**Comment by nostdm** [comment]
I reviewed the latest changes, I think it's good that we remove the unnecessary code from the proxy.

File: contracts/internals/transferrable.sol (hunk 7 above)
Comment (i-stam) [reply contracts/internals/transferrable.sol:37]: Yes, the problem is that due to the extra code added in the fallback function (delegate call etc.) ALL deposits made via
transfer() or send() are broken! At least it is good to offer users the possibility to send funds to other wallets without having to  use executeTransaction(). The problem still remains: if someone is trying to send funds through a contract (e.g. exchanges and they haven't updated they transfer function, the Tx will fail.
